scalding-parquet-scrooge/src/main/scala/com/twitter/scalding/parquet/scrooge/ParquetCollectionFormatCompatibility.scala

```diff
@@ -0,0 +1,163 @@
+        val projectedReadGroupType = projectedReadType.asGroupType
+        val fileGroupType = fileType.asGroupType
+        val projectedReadFields = projectedReadGroupType.getFields.asScala.map { projectedReadField =>
+          if (!fileGroupType.containsField(projectedReadField.getName)) {
+            if (!projectedReadField.isRepetition(Repetition.OPTIONAL)) {
+              throw new DecodingSchemaMismatchException(
+                s"Found non-optional projected read field ${projectedReadField.getName}:\n$projectedReadField\n\n" +
+                  s"not present in the given file group type:\n${fileGroupType}"
+              )
+            }
+            projectedReadField
+          } else {
+            val fileFieldIndex = fileGroupType.getFieldIndex(projectedReadField.getName)
+            val fileField = fileGroupType.getFields.get(fileFieldIndex)
+            if (fileField.isRepetition(Repetition.OPTIONAL) && projectedReadField.isRepetition(Repetition.REQUIRED)) {
+              throw new DecodingSchemaMismatchException(
+                s"Found required projected read field ${projectedReadField.getName}:\n$projectedReadField\n\n" +
+                  s"on optional file field:\n${fileField}"
+              )
+            }
+            projectFileType(fileField, projectedReadField, FieldContext(projectedReadField.getName))
+          }
+        }
+        projectedReadGroupType.withNewFields(projectedReadFields.asJava)
+    }
+  }
+
+  private def projectFileGroup(fileGroup: CollectionGroup,
+                               projectedReadGroup: CollectionGroup,
+                               fieldContext: FieldContext,
+                               formatter: ParquetCollectionFormatter) = {
+    val projectedFileRepeatedType = formatter.formatCompatibleRepeatedType(
+      fileGroup.repeatedType,
+      projectedReadGroup.repeatedType,
+      fieldContext,
+      projectFileType
+    )
+    // Respect optional/required from the projected read group.
+    projectedReadGroup.groupType.withNewFields(projectedFileRepeatedType)
+  }
+
+  private def extractCollectionGroup(typ: Type): Option[CollectionGroup] = {
+    ParquetListFormatter.extractGroup(typ).orElse(ParquetMapFormatter.extractGroup(typ))
+  }
+}
+
+private[scrooge] trait ParquetCollectionFormatter {
+  /**
+   * Format source repeated type in the structure of target repeated type.
+   *
+   * @param readRepeatedType repeated type from which the formatted result get content
+   * @param fileRepeatedType repeated type from which the formatted result get the structure
+   * @param recursiveSolver solver for the inner content of the repeated type
+   * @return formatted result
+   */
+  def formatCompatibleRepeatedType(fileRepeatedType: Type,
+                                   readRepeatedType: Type,
+                                   fieldContext: FieldContext,
+                                   recursiveSolver: (Type, Type, FieldContext) => Type): Type
```

Comment:
In practice, I think this is always invoked with the `projectFileType` method defined above?

Reply:
Yup it is. My thought is to make this recursive call explicit from the caller side so it's clear from the reading within this format compatbility layer instead of jumping into the respective formatter which then cyclically refers back to the caller's method.